src/lib.rs

```diff
@@ -159,4 +160,22 @@ pub trait ServerChannel<In: RpcMessage, Out: RpcMessage, T: ChannelTypes>:
 {
     /// Accept a bidirectional stream
     fn accept_bi(&self) -> T::AcceptBiFuture<'_, In, Out>;
+
+    /// The local addresses this server is bound to.
+    ///
+    /// This is useful for publicly facing addresses when you start the server with a random
+    /// port, `:0` and let the kernel choose the real bind address.  This will return the
+    /// address with the actual port used.
+    fn local_addr(&self) -> &[LocalAddr];
```

Comment:
I'm not really happy with this return type, it was just something to get this working quickly.  I think ideally it would be an iterator so that it can be constructed on the fly, e.g. right now for the mem channel this has to store something on the struct, which is pretty unfortunate.  But because this is a trait I'd have to define an concrete iterator type for that.

Do you think it is worth creating an iterator type?

It is a sequence because of combined btw.

Reply:
So this means that a local addr must be either socket or mem. If you want to make addresses part of this crate (so far they are not modeled), the consistent way would be to have a Addr type on crate::ChannelTypes.

I am not completely sure if we should model it at all in the trait. Can't we just have a way to get the local addr on the concrete structs for which it makes sense?

Reply:
We use this in iroh via `combined` and that almost forces you to do it generically - hence it ended up in the trait.

I understand the reluctance of starting to add addresses, I was only thinking of them as a reporting feature but I can see how that forces you to think more about where else they could end up being used.